bluetooth: host: add ability to load CCC settings on demand

fnde-ot opened the pull request:
This commits adds a BT_SETTINGS_CCC_LAZY_LOADING option to allow for CCC settings to be loaded on demand when a peer device connects in order to reduce memory usage.

fnde-ot committed:
bluetooth: host: add ability to load CCC settings on demand

This commits adds a BT_SETTINGS_CCC_LAZY_LOADING option to allow for
CCC settings to be loaded on demand when a peer device connects in
order to reduce memory usage.

Signed-off-by: François Delawarde <fnde@oticon.com>

joerchan commented:
@fnde-ot Is there any advantage in keeping the old method? Maybe we could skip kconfig option and always do this?

fnde-ot commented:
> @fnde-ot Is there any advantage in keeping the old method? Maybe we could skip kconfig option and always do this?

I think it could end up being very inefficient with some settings backends that have to go through the entire file to find a specific setting. Maybe we could have input from @nvlsianpu!

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -278,13 +278,88 @@ static void sc_clear(struct gatt_sc_cfg *cfg)
-	if (value != BT_GATT_CCC_INDICATE) {
+	if (value == BT_GATT_CCC_INDICATE) {
+		/* Create a new SC configuration entry if subscribed */
+		sc_save(conn->id, &conn->le.dst, 0, 0);
```
Not following the reasoning here, so is this keeping the sc in memory for every paired device? Isn't that what we are trying to get rid of so we only consume ram for connected devices?

fnde-ot replied:
When a database change occurs, we need to know which paired devices are subscribed to SC so we can save the SC data to be used next time the device reconnects. If we only have access to CCC data of connected devices, we need a way to know which devices are subscribed, so we can just keep an SC entry for that purpose.

It uses way less memory to only keep SC in memory rather than keeping data for all CCC attributes.

Vudentz commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -173,6 +173,15 @@ config BT_SETTINGS
 
 if BT_SETTINGS
+config BT_SETTINGS_CCC_LOAD_ON_DEMAND
+	bool "Load CCC values from settings when peer connects"
```
I would default this y since normally devices have less ram then flash.

fnde-ot replied:
I don't mind either ways, it might depend on how fast the setting backend can return a single value. If our existing backends have to load and parse the entire file to find a single setting, some might prefer to keep it disabled by default.
It would be nice to have @nvlsianpu 's opinion on how efficient existing backends are to load single settings before enabling the option by default.

fnde-ot replied:
Will do and update the PR as per @rakons 's comment below.

nvlsianpu replied:
I clarifies my position in the main conversation. This is OK now here.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3594,6 +3599,26 @@ void bt_gatt_connected(struct bt_conn *conn)
 	data.conn = conn;
 	data.sec = BT_SECURITY_L1;
 
+#if defined(CONFIG_BT_SETTINGS_CCC_LOAD_ON_DEMAND)
```
Can't we use IS_ENABLE in this block?

fnde-ot replied:
Not without making a certain amount of unrelated changes, it references ccc_set_direct which references ccc_set which is only defined when BT_SETTINGS is on.

I wanted to keep the PR small, but I think it can be done by moving things around (ccc_set can be moved out of the #if defined(CONFIG_BT_SETTINGS)) if you prefer!

fnde-ot replied:
Addressed in latest push, I separated functionality and moving functions around in separate commits as suggested by @joerchan.

rakons commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4124,7 +4149,27 @@ static int ccc_set(const char *name, size_t len_rd, settings_read_cb read_cb,
+
+	BT_DBG("key: %s", log_strdup((const char *)param));
+
+	/* Only "bt/ccc" settings should ever come here */
```
Do we really need to check it? Maybe assertion would be better - if we get other settings name it seems like an internal code error, isn't it?

fnde-ot replied:
More than a just a check, this is mostly meant to extract the const char *name that is needed by the ccc_set function (it is meant to avoid re-implementing what ccc_set already does in ccc_set_direct).

rakons replied:
Right. Ok then.

rakons reviewed (commented):
I have some troubles with that code. First of all - I find it very usefull, this is something I wished to be implemented but left it to finish other stuff first, so I am glad that you took care of it.

The problems that I have:
1. I can see that CCC is loaded - when it is saved?
2. If we remove bond - what happens with CCC saved?

fnde-ot commented:
>     1. I can see that CCC is loaded - when it is saved?

This commit does not change the way CCC settings are saved, it works the same as it used to (the settings are either saved when a device disconnects, or after a small delay if CONFIG_BT_SETTINGS_CCC_STORE_ON_WRITE is selected)

>     2. If we remove bond - what happens with CCC saved?

This also does not change in this commit, bt_gatt_clear will be called when unpairing a device or when removing a local identity, which in turn will call bt_gatt_clear_ccc.

fnde-ot commented on include/bluetooth/gatt.h:
```diff
@@ -551,7 +551,11 @@ ssize_t bt_gatt_attr_read_chrc(struct bt_conn *conn,
 	BT_GATT_ATTRIBUTE(_uuid, _perm, _read, _write, _value)
 
-#define BT_GATT_CCC_MAX (CONFIG_BT_MAX_PAIRED + CONFIG_BT_MAX_CONN)
+#if defined(CONFIG_BT_SETTINGS_CCC_LOAD_ON_DEMAND)
```
I that a new convention we should follow? I see #if defined(CONFIG_...) used thousands of times in Zephyr, whereas #if IS_ENABLED(CONFIG_...) is used less than 20 times.

rakons replied:
https://github.com/zephyrproject-rtos/zephyr/blob/819276e082e569228de3c5afb140a313df9c0e6c/scripts/checkpatch.pl#L6100

fnde-ot replied:
Good to know, thanks! Will update the PR.

fnde-ot commented:
@rakons Thanks for looking into this, do you think we should enable this feature by default? Will loading single setting whenever a device connects work efficiently with existing backends (FCB, FILE and NVS)?

rakons commented:
> Will loading single setting whenever a device connects work efficiently with existing backends (FCB, FILE and NVS)?

I think it might be enabled by default. In the end - how offten do you connect to a new device? In very specific devices that connects and disconnects a lot, you may wish to disable this feature.

fnde-ot commented:
Note: Just pushed an update that changes #if defined to #if IS_ENABLED, and enables the Kconfig option by default.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3564,6 +3599,26 @@ void bt_gatt_connected(struct bt_conn *conn)
 
+#if IS_ENABLED(CONFIG_BT_SETTINGS_CCC_LOAD_ON_DEMAND)
+	/* Load CCC settings from backend if bonded */
+	if (bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
```
Can't the IS_ENABLED() here be embedded into the if-statement? I.e.
```
if (IS_ENABLED(..) && bt_addr_le_is_bonded(..)) {
```

fnde-ot replied:
Not easily without moving things around (it references ccc_set_direct with references ccc_set which is inside an #if defined), see also above answer to @Vudentz 's comment.

fnde-ot replied:
Addressed in latest push, I separated functionality and moving functions around in separate commits as suggested by @joerchan.

aescolar commented:
@joerchan @jhedberg @Vudentz @nvlsianpu Just a time check, as we are approaching freeze tomorrow

nvlsianpu commented:
@joerchan 

> @fnde-ot Is there any advantage in keeping the old method? Maybe we could skip kconfig option and always do this?

> I think it could end up being very inefficient with some settings backends that have to go through the entire file to find a specific setting. Maybe we could have input from @nvlsianpu!

Unfortunately this will be case for FCB-backend, for NVS-backend should be less affected, but still need to process up to all entries.

What i see here is need for NVS record cache for reference a few nvs-entry needed to be accessed fast. << @Laczen FYI

I would no bother about FCB - NVS is better in all respect

This issue shouldn't be a case for FS-backend (just normal FS access processing addition).

nvlsianpu commented:
^^ worth to emphasize the fact that if a settings constrained beckend is used (either FCB or NVS) most probably it uses quite small flash area, so browsing it should`t be extremely long.

joerchan commented:
@fnde-ot Ok. I see the value in keeping the Kconfig then. Seems sensible to have the option.

joerchan reviewed (commented):
>  Just a time check, as we are approaching freeze tomorrow

@aescolar Not sure I will be able to go through it thoroughly by then, I don't really understand all the changes yet but in general this looks good to me. 
I'm posting my initial comments at least. I have a few questions.

joerchan commented:
@fnde-ot would it possible to make the refactor of moving the static functions as it's own commit. So that it is easier to see the functional change in isolation?

Vudentz commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -173,6 +173,16 @@ config BT_SETTINGS
+	  Load Client Configuration Characteristic value right after bonded
+	  device connects.
+	  By default, CCC is only loaded when calling settings_load.
+	  Choosing this option is will save memory as the CCC values will only
```
Since it it now default to y I guess we should explain the negative effect when disabling it (which would increase memory usage), not the other around.

fnde-ot replied:
Will update.

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1958,7 +1988,7 @@ static void sc_restore(struct bt_conn *conn)
 	}
 
 	/* Reset config data */
-	sc_clear(cfg);
```
It is not a bug fix, just a slight change of how SC settings are handled.

We previously removed the SC setting when restoring (sc_clear). Now we maintain the setting but reset the data to 0 when the SC indication has been sent and only remove the setting when the device unpairs or unsubscribes.

It allows us to keep track of devices that are subscribed to SC without having to loop through CCC data of all bonded devices, because we want CCC data to be loaded on demand when a device connects.

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1671,6 +1685,31 @@ static u8_t notify_cb(const struct bt_gatt_attr *attr, void *user_data)
+			conn = bt_conn_lookup_addr_le(cfg->id,
+						      &cfg->peer);
+			if (!conn) {
+				struct sc_data *sc;
```
It is done 2 lines above (see if (!conn)).

In this callback, we check if the device we want to send the indication to is connected. If it is, we send the indication, otherwise we save the indication data in the SC setting so we can remember to send the indication the next time the device connects.

jhedberg replied:
Yes, finding a connection object doesn't automatically mean that you're connected. There may be a left-over reference but the state is actually disconnected.

jhedberg replied:
There's a `bt_conn_lookup_state_le()` that's probably most convenient here

fnde-ot replied:
Updated to use bt_conn_lookup_state_le() in latest push, to insure we save SC data in the appropriate case.

carlescufi commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -173,6 +173,16 @@ config BT_SETTINGS
 	  instead of each subsystem doing it independently.
 
 if BT_SETTINGS
+config BT_SETTINGS_CCC_LOAD_ON_DEMAND
```
@fnde-ot if it's not too late for the feature freeze today, can we rename this to:

`BT_SETTINGS_CCC_LAZY_LOADING` ?

fnde-ot replied:
I don't mind if you prefer. Just to clarify, would "lazy loading" mean that we load everything at once (lazy because we do it only once at init time) or to load when a device connects (on demand)?

fnde-ot replied:
Just pushed this change!

fnde-ot committed:
bluetooth: host: refactor ccc settings handling

Move functions in order to avoid function prototype and use IS_ENABLED
instead of #if defined where possible.

Signed-off-by: François Delawarde <fnde@oticon.com>

fnde-ot committed:
bluetooth: host: do not remove SC entry when peer subscribed and paired

Maintain Service Changed entry as long as the peer device is subscribed
to SC indications and bonded. This allows to save indication data for
disconnected peers peers when CCC settings are not available (loaded
on-demand).

Signed-off-by: François Delawarde <fnde@oticon.com>

fnde-ot committed:
bluetooth: host: refactor sc settings handling

Move functions in order to avoid function prototypes.

Signed-off-by: François Delawarde <fnde@oticon.com>

fnde-ot commented:
> @fnde-ot would it possible to make the refactor of moving the static functions as it's own commit. So that it is easier to see the functional change in isolation?

I did that in latest push, hopefully it makes it easier to review!

Vudentz reviewed (approved):
LGTM

joerchan reviewed (approved):
Looks good.

I still have a question, but that is not related to the changes in this PR.
